README.md
# Old Weather

The fifth incarnation of OW, built as a custom Angular app. This is a little different from other Zoo apps as it uses a `Makefile` along with Gulp to carry out build / deployment tasks.

## Requirements

- Node / NPM
- Bower
  * `npm install -g bower`
- s3cmd
  * `brew install s3cmd` if on OS X

## Installation

1. Clone the repo
2. Run `make install` inside the repo directory

## Development

Running `make run` starts the project at `http://localhost:8000/`, and the ngDoc server at `http://localhost:8001/`. There are git hooks to run JSLint on commit.

## Deployment

- Requires the environment variables `AMAZON_ACCESS_KEY_ID` and `AMAZON_SECRET_ACCESS_KEY`

Running `make deploy-preview` deploys both project and docs to `http://demo.zooniverse.org/oldweather/` and `http://demo.zooniverse.org/oldweatherdocs/` respectively.

Running `make deploy-production` deploys the project only to `http://www.oldweather.org`

- If you get errors when trying to deploy and you haven't run the dev `make run` cmd, then run the default gulp task to build the required files.
  - `gulp`

## Adding new subject sets

1. Upload your new set to Panoptes
1. Associate the new subject set with the `Annotation` workflow in the workflow editor
1. Modify your subject set metadata to include the following data (both strings):
    ```
    metadata: {
        active: "true",
        shortName: "SHIP_NAME"
    }
    ```
1. Add any extra info to `ShipsDetailConstants` as required

## Gotchas / weirdnesses

- The build task _doesn't_ rebuild the project templates, so if you're just looking to do a simple HTML change, you'll need to e.g. run the dev server for those changes to take effect.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

### Added
- Logout link to nav
- Ship info on the USS Farragut, instructions on adding new sets

### Changed
- Hidden 'Toggle Guides' button until we can work out what it does

### Fixed
- Stopped padding being overridden on home content
- Correct links on ship info page

## [0.1.0] - 14-08-2015
### Added
- Added ship info page, and supporting constants file / styling

### Changed
- Swapped Clarendon for Domine (Google Font, for simples)
- Updated docs and README, log etc

## [0.0.2] - 03-07-2015
### Added
- Misc features and bug fixes. Check the commit log up to 598c9339f680fada67814b8b563c3925d9e5188a

## [0.0.1] - 19-03-2015
### Added
- Initial release, everything up to 78d0520169cfa456c977d729020f2066ad0e27a8




gulpfile.js
var fs = require('fs');

var gulp = require('gulp');
var gulpDocs = require('gulp-ngdocs');
var runSequence = require('run-sequence');
var chmod = require('gulp-chmod');
var watch = require('gulp-watch');
var clean = require('gulp-clean');
var connect = require('gulp-connect');
var sourcemaps = require('gulp-sourcemaps');
var stylus = require('gulp-stylus');
var nib = require('nib');
var templateCache = require('gulp-angular-templatecache');

var usemin = require('gulp-usemin');
var ngAnnotate = require('gulp-ng-annotate');
var uglify = require('gulp-uglify');
var minifyHTML = require('gulp-minify-html');
var minifyCSS = require('gulp-minify-css');
var rev = require('gulp-rev');
var karma = require('karma').server;
var jshint = require('gulp-jshint');

var baseDir = __dirname
var appDir = baseDir + '/app';
var modulesDir = appDir + '/modules';
var stylDir = baseDir + '/styl';
var files = {
    templates: appDir + '/modules/**/templates/**/*.html',
    scripts: modulesDir + '/**/*.js'
}

gulp.task('cleanHooks', function () {
    return gulp.src('.git/hooks/pre-commit', {read: false}).pipe(clean());
});

gulp.task('cleanTemplates', function () {
    return gulp.src('**/templates.js', {read: false}).pipe(clean());
});

gulp.task('cleanBuild', function () {
    return gulp.src('.tmp/build', {read: false}).pipe(clean());
});

gulp.task('cleanDocs', function () {
    return gulp.src('.tmp/docs', {read: false}).pipe(clean());
});

gulp.task('stylus', function () {
    gulp.src(stylDir + '/main.styl')
        .pipe(sourcemaps.init())
        .pipe(stylus({use: nib()}))
        .pipe(sourcemaps.write())
        .pipe(gulp.dest(appDir + '/css'));
});

gulp.task('connectDev', function () {
    connect.server({
        root: ['app'],
        port: 8000
    });
});

gulp.task('connectDocs', function () {
    connect.server({
        root: ['.tmp/docs'],
        port: 8001
    });
});

gulp.task('makeHooks', function () {
    gulp.src('git-hooks/pre-commit')
        .pipe(chmod(755))
        .pipe(gulp.dest('.git/hooks/'));
});

gulp.task('jshint', function () {
    gulp.src(modulesDir + '/**/!(templates).js')
        .pipe(jshint('.jshintrc'))
        .pipe(jshint.reporter('jshint-stylish'));
});

gulp.task('karma', function (cb) {
    karma.start({
        configFile: __dirname + '/karma.conf.js'
    }, cb);
});

gulp.task('karma-ci', function (cb) {
    karma.start({
        configFile: __dirname + '/karma.conf.js',
        singleRun: true,
        coverageReporter: {
            type: 'lcovonly',
            // Travis uses this path: coverage/lcov.info
            subdir: '.',
            file: 'lcov.info'
        }
    }, cb);

});

gulp.task('coffee', function () {
    gulp.src(files.coffee)
    .pipe(sourcemaps.init())
    .pipe(coffee().on('error', gutil.log))
    .pipe(sourcemaps.write())
    .pipe(ngAnnotate())
    .pipe(gulp.dest(modulesDir + '/'))
});

gulp.task('usemin', function () {
    return gulp.src(appDir + '/index.html')
    .pipe(usemin({
        css: [minifyCSS(), 'concat', rev()],
        html: [minifyHTML({empty: true})],
        js: [ngAnnotate(), uglify(), rev()]
    }))
    .pipe(gulp.dest('.tmp/build/'));
});

gulp.task('templates', function () {
    fs.readdir(modulesDir, function (err, dirs) {
        dirs.forEach(function (dir) {
            var path = modulesDir + '/' + dir;
            gulp.src(path + '/**/templates/**/*.html')
                .pipe(minifyHTML())
                .pipe(templateCache({
                    module: dir,
                    templateHeader: '(function (angular) {\n "use strict";\n angular.module("<%= module %>"<%= standalone %>).run(["$templateCache", function($templateCache) {',
                    templateFooter: '\n }]);\n }(window.angular));'
                }))
                .pipe(gulp.dest(path + '/scripts'));
        });
    });
});

gulp.task('copy', function () {
    gulp.src(appDir + '/images/**/*')
        .pipe(gulp.dest('.tmp/build/images'));

    gulp.src(appDir + '/bower_components/bootstrap/dist/fonts/*')
        .pipe(gulp.dest('.tmp/build/fonts'));
});

gulp.task('ngdocs', function () {
    var options = {
        html5Mode: false,
        title: 'Old Weather Documentation'
    };

    gulp.src(files.scripts)
        .pipe(gulpDocs.process(options))
        .pipe(gulp.dest('.tmp/docs'));
});

gulp.task('watch', function () {
    gulp.watch([files.templates], ['templates']);
    gulp.watch([files.scripts], ['ngdocs']);
    gulp.watch([stylDir + '/**/*'], ['stylus']);
});

gulp.task('default', function (cb) {
    runSequence(
        'cleanDocs',
        'cleanTemplates',
        'templates',
        'ngdocs',
        'stylus',
        [
            'watch',
            'connectDev',
            'connectDocs'
        ],
        cb
    )
});

gulp.task('docs', function (cb) {
    runSequence(
        'cleanDocs',
        'ngdocs',
        cb
    )
});

gulp.task('build', function (cb) {
    runSequence(
        'cleanBuild',
        'copy',
        'usemin',
        cb
    );
});

gulp.task('test', function (cb) {
    runSequence(
        'jshint',
        'karma',
        cb
    )
});

gulp.task('test-ci', function (cb) {
    runSequence(
        'jshint',
        'karma-ci',
        cb
    )
});

gulp.task('hookmeup', function (cb) {
    runSequence(
        'cleanHooks',
        'makeHooks',
        cb
    );
});
